### PR TITLE
Add RIC method for sparse match interpolation

### PR DESCRIPTION
**Pull request by tsenst**
* commit implementation of RIC interpolation method (from "Robust Interpolation of Correspondences for Large Displacement Optical Flow" Yinlin et. al 2017)
* integrate RIC interpolation to DenseRLOFOpticalFlow class and functions
* add RLOF_EPIC and RLOF_RIC to (sample)optical_flow_estimation project
* implement interface to allow appling better edge estimators (e.g. SED) for edge-awar cost map computation that is used in the EdgeAwareInterpolator and RICInterpolator

### CODE REVIEW

File: modules/ximgproc/CMakeLists.txt [hunk 1]
```diff
@@ -1,2 +1,2 @@
 set(the_description "Extended image processing module. It includes edge-aware filters and etc.")
-ocv_define_module(ximgproc opencv_core opencv_imgproc opencv_calib3d opencv_imgcodecs WRAP python java)
+ocv_define_module(ximgproc opencv_core opencv_imgproc opencv_calib3d opencv_imgcodecs opencv_tracking  WRAP python java)
```
Comment (alalek) [inline modules/ximgproc/CMakeLists.txt]: We prefer to keep dependency list minimal - adding tracking here, this would block adding ximgproc as dependency into tracing module.

Do we really need this dependency?

File: modules/ximgproc/src/sparse_match_interpolators.cpp [hunk 2]
```diff
@@ -36,8 +36,10 @@
 
 #include "precomp.hpp"
 #include "opencv2/ximgproc/sparse_match_interpolator.hpp"
+#include "opencv2/video.hpp"
```
Comment (alalek) [inline modules/ximgproc/src/sparse_match_interpolators.cpp]: Module's dependency list should know that.
Please avoid indirect dependencies.

File: modules/ximgproc/src/sparse_match_interpolators.cpp [hunk 3]
```diff
@@ -36,8 +36,10 @@
...
 #include <algorithm>
 #include <vector>
+#include <iostream>
```
Comment (alalek) [inline modules/ximgproc/src/sparse_match_interpolators.cpp]: used?

File: modules/ximgproc/include/opencv2/ximgproc/sparse_match_interpolator.hpp [hunk 4]
```diff
@@ -125,6 +137,132 @@ EdgeAwareInterpolator.
...
+public:
+    /** @brief K is a number of nearest-neighbor matches considered, when fitting a locally affine
+     *model for a superpixel segment. However, lower values would make the interpolation
+     *noticeably faster. The original implementation of @cite Hu2017 uses 32.
+    */
+    CV_WRAP virtual void setK(int _k = 32) = 0;
+    /** @copybrief setK
+     *    @see setK
+     */
+    CV_WRAP virtual int  getK() const = 0;
```
Comment (alalek) [inline modules/ximgproc/include/opencv2/ximgproc/sparse_match_interpolator.hpp]: please remove extra double spaces here and below.

File: modules/ximgproc/include/opencv2/ximgproc/sparse_match_interpolator.hpp [hunk 5]
```diff
@@ -125,6 +137,132 @@ EdgeAwareInterpolator.
...
+    /** @brief Interface to provide a more elaborated cost map, i.e. edge map, for the edge-aware term.
+     *  This implementation is based on a rather simple gradient-based edge map estimation.
+     *  To used more complex edge map estimator (e.g. StructuredEdgeDetection that has been
+     *  used in the original publication) that may lead to improved accuracies, the internal
+     *  edge map estimation can be bypassed here.
+     *  @param _costMap a type CV_32FC1 Mat is required.
+     *  @see cv::ximgproc::createSuperpixelSLIC
+    */
+    CV_WRAP virtual void setCostMap(const Mat & _costMap) = 0;
```
Comment (alalek) [inline modules/ximgproc/include/opencv2/ximgproc/sparse_match_interpolator.hpp]: > _costMap => costMap

Please avoid using of underscored parameters in .hpp files (there is no conflict with fields/variables). These names are visible in documentation - lets keep them clean.

File: modules/optflow/doc/optflow.bib [hunk 6]
```diff
@@ -104,9 +104,26 @@ @article{Senst2012
   number = {9},
 }
 
+@phdthesis{Senst2019,
+  title={Estimation and analysis of motion in video data},
+  author={Senst, Tobias},
+  year={2019},
+  school={Technical Univerity Berlin},
+   doi = {10.14279/depositonce-9085},
```
Comment (alalek) [inline modules/optflow/doc/optflow.bib]: indentation

File: modules/ximgproc/src/sparse_match_interpolators.cpp [hunk 7]
```diff
@@ -201,13 +207,13 @@ void EdgeAwareInterpolatorImpl::interpolate(InputArray from_image, InputArray fr
     if(use_post_proc)
         fastGlobalSmootherFilter(src,dst,dst,fgs_lambda,fgs_sigma);
 
+    costMap = Mat();
```
Comment (alalek) [inline modules/ximgproc/src/sparse_match_interpolators.cpp]: `costMap.release();` should do the same.

File: modules/ximgproc/src/sparse_match_interpolators.cpp [hunk 8]
```diff
@@ -877,5 +888,1148 @@ bool operator<(const SparseMatch& lhs,const SparseMatch& rhs)
...
+    bool refine_models;
+    bool use_variational_refinement;
+    float cost_suffix;
+    float max_flow;
+    bool use_global_smoother_filter;
+    float fgs_lambda;
+    float fgs_sigma;
+    SLICType slic_type;
+
+    class GetKNNMatches_ParBody : public ParallelLoopBody
```
Comment (alalek) [inline modules/ximgproc/src/sparse_match_interpolators.cpp]: BTW, "master" branch allows using of C++11 features, including lambdas for `parallel_for_()`.

It is not necessary to move all code into caller functions, you might call methods from lambdas (with extra parameters if needed).

File: modules/ximgproc/src/sparse_match_interpolators.cpp [hunk 9]
```diff
@@ -877,5 +888,1148 @@ bool operator<(const SparseMatch& lhs,const SparseMatch& rhs)
...
+    int m_validSize;
+};
+
+void RICInterpolatorImpl::interpolate(InputArray from_image, InputArray from_points, InputArray to_image, InputArray to_points, OutputArray dense_flow)
+{
+    CV_Assert(!from_image.empty() && (from_image.depth() == CV_8U) && (from_image.channels() == 3 || from_image.channels() == 1));
+    CV_Assert( use_variational_refinement == false || (!to_image.empty() && (to_image.depth() == CV_8U) && (to_image.channels() == 3 || to_image.channels() == 1)));
+    CV_Assert(!from_points.empty() && from_points.isVector() &&
+        !to_points.empty() && to_points.isVector() &&
+        from_points.sameSize(to_points));
```
Comment (alalek) [inline modules/ximgproc/src/sparse_match_interpolators.cpp]: Avoid using of complex assertion statements - they emits over-complicated unclear messages to Users.

```
CV_Assert(!from_points.empty() && from_points.isVector());
CV_Assert(!to_points.empty() && to_points.isVector());
CV_Assert(from_points.sameSize(to_points));
```

Their messages helping to localize parameters mistakes much better.

File: modules/ximgproc/src/sparse_match_interpolators.cpp [hunk 10]
```diff
@@ -877,5 +888,1148 @@ bool operator<(const SparseMatch& lhs,const SparseMatch& rhs)
...
+    {
+        m_validSize = 0;
+    }
+    int Size()
+    {
+        return m_validSize;
+    }
+
+private:
+    inline void swap(float & a, float & b)
```
Comment (alalek) [inline modules/ximgproc/src/sparse_match_interpolators.cpp]: Why is not `std::swap()`?

File: modules/ximgproc/src/sparse_match_interpolators.cpp [hunk 11]
```diff
@@ -877,5 +888,1148 @@ bool operator<(const SparseMatch& lhs,const SparseMatch& rhs)
...
+        int startPos = 0, endPos = spCnt, step = 1;
+        if (iter % 2 == 1) {
+            startPos = spCnt - 1; endPos = -1; step = -1;
+        }
+
+        for (int idx = startPos; idx != endPos; idx += step) {
+
+            //if (idx >= 3 * (endPos + startPos) / 4) {
+                //break;
+            //}
```
Comment (alalek) [inline modules/ximgproc/src/sparse_match_interpolators.cpp]: dead code?

File: modules/ximgproc/CMakeLists.txt (hunk 1 above)
Comment (tsenst) [reply modules/ximgproc/CMakeLists.txt]: Thank you for that comment. In fact tracking is not needed anymore.

File: modules/ximgproc/src/sparse_match_interpolators.cpp (hunk 2 above)
Comment (tsenst) [reply modules/ximgproc/src/sparse_match_interpolators.cpp]: video.hpp is required for the application of the VariationalRefinement method. However even if I add opencv_video to the ocv_define_module definitions. Is the a better way avoid this indirect dependency?

File: modules/ximgproc/src/sparse_match_interpolators.cpp (hunk 8 above)
Comment (tsenst) [reply modules/ximgproc/src/sparse_match_interpolators.cpp]: Can you give me a link to an function in the OpenCV lib where this has been used. So I can see how lambdas in that case are used correctly?

File: modules/ximgproc/src/sparse_match_interpolators.cpp (hunk 11 above)
Comment (tsenst) [reply modules/ximgproc/src/sparse_match_interpolators.cpp]: has been removed.

File: modules/ximgproc/src/sparse_match_interpolators.cpp (hunk 10 above)
Comment (tsenst) [reply modules/ximgproc/src/sparse_match_interpolators.cpp]: swap function has been removed

File: modules/ximgproc/src/sparse_match_interpolators.cpp (hunk 9 above)
Comment (tsenst) [reply modules/ximgproc/src/sparse_match_interpolators.cpp]: done

File: modules/ximgproc/src/sparse_match_interpolators.cpp (hunk 8 above)
Comment (tsenst) [reply modules/ximgproc/src/sparse_match_interpolators.cpp]: done

File: modules/ximgproc/src/sparse_match_interpolators.cpp (hunk 7 above)
Comment (tsenst) [reply modules/ximgproc/src/sparse_match_interpolators.cpp]: done

File: modules/ximgproc/include/opencv2/ximgproc/sparse_match_interpolator.hpp (hunk 5 above)
Comment (tsenst) [reply modules/ximgproc/include/opencv2/ximgproc/sparse_match_interpolator.hpp]: Underscores have been removed

File: modules/ximgproc/include/opencv2/ximgproc/sparse_match_interpolator.hpp (hunk 4 above)
Comment (tsenst) [reply modules/ximgproc/include/opencv2/ximgproc/sparse_match_interpolator.hpp]: Double spaces have been removed

File: modules/ximgproc/src/sparse_match_interpolators.cpp (hunk 3 above)
Comment (tsenst) [reply modules/ximgproc/src/sparse_match_interpolators.cpp]: Removed

File: modules/ximgproc/src/sparse_match_interpolators.cpp (hunk 7 above)
Comment (alalek) [reply modules/ximgproc/src/sparse_match_interpolators.cpp]: where?

File: modules/ximgproc/src/sparse_match_interpolators.cpp [hunk 12]
```diff
@@ -877,5 +887,1026 @@ bool operator<(const SparseMatch& lhs,const SparseMatch& rhs)
...
+    g = new vector<node>[match_num];
+    buildGraph(matDistanceMap, costMap);
+    parallel_for_(Range(0, getNumThreads()), [&](const Range & range)
+    {
+        int stripe_sz = (int)ceil(match_num / (double)getNumThreads());
+        int start = std::min(range.start * stripe_sz, match_num);
+        int end = std::min(range.end   * stripe_sz, match_num);
+        nodeHeap q((int)match_num);
+        int num_expanded_vertices;
+        unsigned char* expanded_flag = new unsigned char[match_num];
```
Comment (alalek) [inline modules/ximgproc/src/sparse_match_interpolators.cpp]: Please avoid manual memory management. Use `cv::AutoBuffer<unsigned char>` or `std::vector` instead.

File: modules/ximgproc/src/sparse_match_interpolators.cpp [hunk 13]
```diff
@@ -877,5 +887,1026 @@ bool operator<(const SparseMatch& lhs,const SparseMatch& rhs)
...
+    {
+        const SparseMatch & p = matches_vector[i];
+        Point pos(static_cast<int>(p.reference_image_pos.x), static_cast<int>(p.reference_image_pos.y));
+        labels.at<int>(pos) = i;
+        matDistanceMap.at<float>(pos) = costMap.at<float>(pos);
+    }
+
+    geodesicDistanceTransform(matDistanceMap, costMap);
+
+    g = new vector<node>[match_num];
```
Comment (alalek) [inline modules/ximgproc/src/sparse_match_interpolators.cpp]: Please avoid manual memory management.

File: modules/ximgproc/src/sparse_match_interpolators.cpp [hunk 14]
```diff
@@ -877,5 +887,1026 @@ bool operator<(const SparseMatch& lhs,const SparseMatch& rhs)
...
+                            bestCost[idx] = cost;
+                        }
+                    }
+                }
+                vFlags[idx] = 1;
+            }
+        }
+    });
+
+    //parallel_for_(Range(0, iterCnt), GenerateHypothesis_ParBody(*this, &bestCost[0], outModels, supportCnt, spCnt, spNN, &supportMatchIds[0], &supportMatchDis[0], inputMatches, inLierFlag));
```
Comment (alalek) [inline modules/ximgproc/src/sparse_match_interpolators.cpp]: dead code?

File: modules/tracking/samples/tracker.py [hunk 15]
```diff
@@ -2,13 +2,15 @@
 import cv2 as cv
 import sys
 
-if len(sys.argv) != 2:
-    print('Input video name is missing')
-    exit()
+#if len(sys.argv) != 2:
+#    print('Input video name is missing')
+#    exit()
```
Comment (alalek) [inline modules/tracking/samples/tracker.py]: please rollback unrelated changes here and below

File: modules/ximgproc/include/opencv2/ximgproc/sparse_match_interpolator.hpp [hunk 16]
```diff
@@ -125,6 +137,132 @@ EdgeAwareInterpolator.
...
+    */
+    CV_WRAP virtual int getSuperpixelNNCnt() const = 0;
+    /** @brief Parameter to tune enforcement of superpixel smoothness factor used for oversegmentation.
+     *  @see cv::ximgproc::createSuperpixelSLIC
+    */
+    CV_WRAP virtual void setSuperpixelRuler(float ruler = 15.f) = 0;
+    /** @copybrief setSuperpixelRuler
+     *    @see setSuperpixelRuler
+     */
+    CV_WRAP virtual float  getSuperpixelRuler() const = 0;
```
Comment (alalek) [inline modules/ximgproc/include/opencv2/ximgproc/sparse_match_interpolator.hpp]: please remove double space

File: modules/ximgproc/src/sparse_match_interpolators.cpp [hunk 17]
```diff
@@ -57,7 +58,8 @@ struct SparseMatch
 
 bool operator<(const SparseMatch& lhs,const SparseMatch& rhs);
 
-void weightedLeastSquaresAffineFit(int* labels, float* weights, int count, float lambda, SparseMatch* matches, Mat& dst);
+void computeGradientMagnitude(Mat& src, Mat& dst);
+void weightedLeastSquaresAffineFit(int* labels, float* weights, int count, float lambda, const SparseMatch* matches, Mat& dst);
```
Comment (alalek) [inline modules/ximgproc/src/sparse_match_interpolators.cpp]: Please declare local (this file only) functions as "static". Or use anonymous namespace for them.

File: modules/ximgproc/src/sparse_match_interpolators.cpp [hunk 18]
```diff
@@ -877,5 +887,1026 @@ bool operator<(const SparseMatch& lhs,const SparseMatch& rhs)
...
+    float getFGSLambda() const CV_OVERRIDE { return fgs_lambda; }
+    void  setFGSSigma(float _sigma) CV_OVERRIDE { fgs_sigma = _sigma; }
+    float getFGSSigma() const CV_OVERRIDE { return fgs_sigma; }
+};
+
+Ptr<RICInterpolatorImpl> RICInterpolatorImpl::create()
+{
+    RICInterpolatorImpl *eai = new RICInterpolatorImpl();
+    eai->init();
+    return Ptr<RICInterpolatorImpl>(eai);
```
Comment (alalek) [inline modules/ximgproc/src/sparse_match_interpolators.cpp]: Memory leak is possible in case of exception in "init".
Don't use raw pointers:
```
auto eai = makePtr<RICInterpolatorImpl>();
eai->init()
return eai;
```

File: modules/ximgproc/src/sparse_match_interpolators.cpp (hunk 7 above)
Comment (tsenst) [reply modules/ximgproc/src/sparse_match_interpolators.cpp]: Oh sorry. Got the wrong line of code in line 1306 was the same issue

File: modules/ximgproc/src/sparse_match_interpolators.cpp (hunk 17 above)
Comment (tsenst) [reply modules/ximgproc/src/sparse_match_interpolators.cpp]: Static has been added.

File: modules/ximgproc/src/sparse_match_interpolators.cpp (hunk 18 above)
Comment (tsenst) [reply modules/ximgproc/src/sparse_match_interpolators.cpp]: Thank you for that hint!

File: modules/ximgproc/src/sparse_match_interpolators.cpp (hunk 7 above)
Comment (alalek) [reply modules/ximgproc/src/sparse_match_interpolators.cpp]: There are two extra cases:
 ` = Mat();` => `.release();`

File: modules/ximgproc/src/sparse_match_interpolators.cpp [hunk 19]
```diff
@@ -877,5 +887,1007 @@ bool operator<(const SparseMatch& lhs,const SparseMatch& rhs)
...
+        m_validSize = 0;
+
+        return 0;
+    }
+    int Push(float index, float weight)
+    {
+        if (m_validSize >= m_size)
+        {
+            CV_Error(CV_StsOutOfRange, " m_validSize >= m_size this problem can be resolved my decreasig k parameter");
+        };
```
[the remark this answers is not in the document]
Comment (tsenst) [reply modules/ximgproc/src/sparse_match_interpolators.cpp]: Has been removed

File: modules/ximgproc/src/sparse_match_interpolators.cpp [hunk 20]
```diff
@@ -877,5 +887,1007 @@ bool operator<(const SparseMatch& lhs,const SparseMatch& rhs)
...
+    float c1 = 1.0f / 2.0f;
+    float c2 = sqrt(2.0f) / 2.0f;
+    float d = 0.0f;
+    int i, j;
+    float *dist_row, *cost_row;
+    float *dist_row_prev, *cost_row_prev;
+    int *label_row;
+    int *label_row_prev;
+
+#define CHECK(cur_dist,cur_label,cur_cost,prev_dist,prev_label,prev_cost,coef)\
```
[the remark this answers is not in the document]
Comment (tsenst) [reply modules/ximgproc/src/sparse_match_interpolators.cpp]: change CHECK to CHK_GD

File: modules/ximgproc/src/sparse_match_interpolators.cpp [hunk 21]
```diff
@@ -877,5 +887,1007 @@ bool operator<(const SparseMatch& lhs,const SparseMatch& rhs)
...
+}
+
+void RICInterpolatorImpl::superpixelNeighborConstruction(const Mat & _labels, int labelCnt, Mat& outNeighbor)
+{
+    // init
+    outNeighbor = Mat(labelCnt, max_neighbors, CV_32SC1); // only support 32 neighbors
+    outNeighbor.setTo(-1);
+
+    //
+    vector<int> diffPairs(labels.cols*_labels.rows * 4, 0);
```
[the remark this answers is not in the document]
Comment (tsenst) [reply modules/ximgproc/src/sparse_match_interpolators.cpp]: removed
